src/build_system.ml

```diff
@@ -398,7 +399,15 @@ let compile_rule t ~all_targets_by_dir ?(allow_override=false) pre_rule =
     in
     let targets_min_ts = min_timestamp t targets_as_list  in
     let deps_max_ts    = max_timestamp t all_deps_as_list in
-    if rule_changed ||
+    let is_runtest ps  =
+      let is_runtest p =
```

Comment:
These is a bit weird, the toplevel could be `are_runtest`, but I think it's simpler to just put the `Pset.for_all` directly in the condition